README.md
# Radio

A headless music player which supports internet radio and streaming.

## Getting Started

- Get the latest release from
  [github.com/MikMuellerDev/radio/releases](https://github.com/MikMuellerDev/radio/releases)
- Extract the tar archive

### Installation

- Make sure that `sudo` is installed on the target machine
- Execute the file `install.sh` as non-root
- The application files will be installed at `/usr/bin/radio`
- The configuration file will be installed at `/etc/radio/config.toml`
- Radio can then be controlled via `systemd`:
  `sudo systemctl enable radio --now`

### Without Installation

- Execute the following command to start radio `./radio run`
- On the first launch, a new configuration file will be created without starting
  the service

## Configuration

Under normal conditions, the service creates a configuration file named
`config.toml` in the working directory where it is launched. If the service is
launched with the `-c` flag, the configuration path can be customized.

### Server Configuration

```toml
port = 8083
session_key = "must be over 64 characters long"
```

The configuration file includes the `port` variable which determines the port on
which radio's HTTP server will listen on.

The `session_key` must be over 64 characters long and should be kept secret. On
Unix-like systems, such a key can be generated using the following command:

```bash
openssl rand -hex 64
```

### Adding Users

Users can be added in the `users` list in the configuration file. Every user
must have a unique username and should be protected via a secure password.

```toml
[[users]]
username = "admin" # A unique username
password = "secret" # A secure password

[[users]]
username = "another_username"
password = "secret2"
```

### Adding Radio Stations / Audio Streams

Generally speaking, radio should support most **MP3** network streams. It does
not matter whether the audio is a livestream or a fixed-length file. Streams can
be added in the `stations` list in the configuration file.

```toml
[[stations]]
id = "example" # An arbitrary (unique) ID for the stream
name = "Example Radio" # A user-friendly name
description = "My description" # A user-friendly description
url = "https://example.com/stream" # The MP3 source of the stream
image_file = "example.png" # The image file inside the `image` directory
auto_restart = true # Whether the stream should be restarted if it stops of fails
auto_start = false # Whether the stream should play as soon as the service is launched

[[stations]]
# ...
```

## Screenshots

### Dashboard

![Dashboard](./screenshots/dash.png)

### Login screen

![Login](./screenshots/login.png)


CHANGELOG.md
# Changelog

- This is the first release


src/config.rs
use anyhow::{bail, Ok, Result};
use std::{
    collections::HashSet,
    fs::{self, File},
    io::Write,
    path::{Path, PathBuf},
};

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct Config {
    pub port: u16,
    pub session_key: String,
    pub users: Vec<User>,
    pub stations: Vec<Station>,
}

#[derive(Serialize, Deserialize)]
pub struct User {
    pub username: String,
    pub password: String,
}

#[derive(Serialize, Deserialize, Clone)]
pub struct Station {
    pub id: String,
    pub name: String,
    pub description: String,
    pub url: String,
    pub image_file: PathBuf,
    pub auto_restart: bool,
    pub auto_start: bool,
}

impl Config {
    fn validate(&self) -> Result<()> {
        let key_len = self.session_key.len();
        if key_len < 64 {
            bail!("insufficient key length: key must be >= 64 characters, found {key_len}",)
        }
        if self.stations.is_empty() {
            bail!("no stations configured: there must be at least one stations")
        }

        let mut station_ids = HashSet::new();
        let mut auto_start_id = None;

        for station in &self.stations {
            // check if station ID is `url`
            if station.id == "url" {
                bail!("station ID cannot be `url`")
            }

            // check if station ID is unique
            if !station_ids.insert(&station.id) {
                bail!("duplicate station ID `{} ", station.id)
            }

            if station.auto_start {
                // validate that only one station is marked as `auto_start`
                match auto_start_id {
                Some(old) => bail!("station `{}` cannot be `auto_start`: the station `{old}` is already marked as `auto_start`", station.id),
                None =>  auto_start_id = Some(station.id.to_string()),
            }
            }

            // validate image of the station
            let path = PathBuf::from("./images").join(&station.image_file);
            if !path.exists() {
                bail!(
                    "station `{}` has an invalid image name: `{}`: path does not exist",
                    station.name,
                    station.image_file.to_string_lossy()
                )
            }
            if !path.is_file() {
                bail!(
                    "station `{}` has an invalid image path: `{}`: path is not a file",
                    station.name,
                    station.image_file.to_string_lossy()
                )
            }
        }
        Ok(())
    }
}

pub fn read(path: &Path) -> Result<Option<Config>> {
    match path.exists() {
        true => {
            let raw_config = fs::read_to_string(path)?;
            let config = toml::from_str::<Config>(&raw_config)?;
            config.validate()?;
            Ok(Some(config))
        }
        false => {
            // create the config file using a default
            fs::create_dir_all(path.parent().unwrap())?;
            let mut file = File::create(path)?;

            /* let config = Config {
                port: 8083,
                session_key: "my_secret_session_key_which_must_be_over_64_characters_in_length".to_string(),
                users: vec![User {
                    username: "admin".to_string(),
                    password: "secret".to_string(),
                }],
                stations: vec![Station {
                    id: "example".to_string(),
                    name: "Example Radio".to_string(),
                    description: "This is an example radio".to_string(),
                    url: "https://example.com/stream".to_string(),
                    image_file: PathBuf::from("example.png"),
                    auto_restart: true,
                }],
            };
            file.write_all(toml::to_string(&config).unwrap().as_bytes()); */

            file.write_all(include_bytes!("default_config.toml"))?;
            Ok(None)
        }
    }
}
